warplib/src/warp_memory/arrays.cairo

```diff
@@ -0,0 +1,93 @@
+use warplib::warp_memory::WarpMemory;
+use warplib::warp_memory::WarpMemoryTrait;
+use warplib::warp_memory::WarpMemoryImpl;
+use integer::u256_from_felt252;
+use integer::u128_try_from_felt252;
+use integer::Into;
+use option::OptionTrait;
+
+trait WarpMemoryArraysTrait {
+    /// Reads the pointer of an array. If it does not exist, it creates a new one
+    fn get_or_create_id(ref self: WarpMemory, location: felt252, size: felt252) -> felt252;
+
+    /// Given the length and the felt size of an element, allocates the space
+    /// in memory and returns a pointer to it
+    fn new_dynamic_array(ref self: WarpMemory, len: felt252, elem_width: felt252) -> felt252;
+
+    /// Given a pointer to a dynamic array, the index you want to access from it
+    /// and the size of the element in such array, returns a pointer to the element
+    fn index_dyn(ref self: WarpMemory, array_ptr: felt252, index: felt252, elem_width: felt252) -> felt252;
+
+    /// Given a pointer to a static array, the index you want to access from it,
+    /// the size of the elements in the array and the it's length,  returns a pointer
+    /// to the element
+    fn index_static(ref self: WarpMemory, array_ptr: felt252, index: felt252, elem_width: felt252, length: felt252) -> felt252;
+
+    /// Given the pointer to a dynamic array, it returns it's length
+    fn length_dyn(ref self: WarpMemory, array_ptr: felt252) -> felt252;
+}
+
+
+impl WarpMemoryArraysImpl of WarpMemoryArraysTrait {
+    fn get_or_create_id(ref self: WarpMemory, location: felt252, size: felt252) -> felt252{
+        // Get the id at the location
+        let id = self.read(location);
+        if id != 0 {
+            return id;
+        }
+
+        // If no id was found create a new one
+        let id = self.alloc(size);
+        self.unsafe_write(location, id);
+        id
+    }
+
+    fn new_dynamic_array(ref self: WarpMemory, len: felt252, elem_width: felt252) -> felt252 {
+        // The max space needed for the array in memory
```

Comment:
Why the max possible space if it's a dynamic array?

Reply:
solidity memory arrays work the same as C# arrays. The size can be dynamically set once, after that you cannot change them. A memory dynamic array cannot alter it's size once created

Reply:
Solidity memory arrays are kind of fixed size, just like c# as you said. But we are now at Cairo level and this is a function named `new_dynamic_array` so when i review it I was expecting the creation of an array that can grow and shrink during execution; perhaps the confusion is with that function name and what we are doing here is the creation of a fixed array. What do you think?

Reply:
They are not a fixed size, they have a dynamic size which is set on runtime. Fixed size arrays are static arrays whose size is know during compilation.

Reply:
Or not. We have been referring to them as dynamic for a year now. So the whole code refers them as dynamic and for a rename here to be less confusing, we should then rename many other things to appropriately  refer to this kind of array. 
Also in solidity there is no semantic difference when creating storage arrays vs memory arrays, and the former is truly dynamic.

I think if we rename we should go and rename everything that deals with memory and arrays, and that's a lot of code. I think renaming this single function can cause confusion since this type is treated as dynamic arrays everywhere.

Reply:
I think we at least should do a comment there for futures contributors. I'm in favor of avoid confusions and do a proper naming, I understand it involve some more changes so perhaps could be a good Idea to create a ticket for it